### PR TITLE
Add the combine feature to support multiple chapters

### PR DESCRIPTION
**Pull request by wongpratan**

### CODE REVIEW

**Comment by megahirt** [comment]
Hi @wongpratan , this is a good start.  I've finished reading through your code, and I think I understand what you are doing:

- you add a new bbk CLI command called **bbkCombine** which basically wraps a foreach loop around the **bbkConvert** command, taking as input a text file of paths
- you modify the karaoke.js file to run the combine command

Your change adds over 100 lines of code, which is mostly due to adding this feature as a separate command in the CLI.  The actual feature of this change is contained in a few lines of code:

> shell.exec(
>                 `"${ffmpegExe}" -f concat -safe 0 -i "${Options.list}" -c copy "${Options.output}"`,
> 

where Options.list is a string list of paths (joined together properly)

I would like to suggest a simpler approach:

Integrate the above `shell.exec()` directly into the bbkConvert command, with the following features:

- bbkConvert should accept and make use of a list of paths.  This appears to already be true, however the list is never used (only the first item is used).  See line 124: 
> Options.pathFolder = options._.shift();

- rename Options.pathFolder to Options.pathFolders (plural)
- if pathFolders contains one item, just convert it as normal
- add a command line flag, something like --combineOutput which is a boolean.  If this flag is true, then it will determine what happens next
- if pathFolders contains more than one item:
     - if --combineOutput is set to true, then use the ffmpeg command which you wrote (quoted above) to combine them
     - if --combineOutput is set to false, simply loop over the pathFolders and call ffmpeg as if before.

Does this make sense?  In this way, we will add both features (combine into single file + output multiple files) in a lot less code / complexity.  Let me know if you have questions.

**Comment by wongpratan** [comment]
Thank your @megahirt your review and suggestion.
I'm updating **convert.js** follow your suggestion.
`bbkConvert` has `output` parameter. It is name of a output file.

In case of `--combineOutput` is set to **false**, then loop over the pathFolders.
Should it  generate many output files ? Should we change `output` parameter to be array ?

**Comment by megahirt** [comment]
@wongpratan my apologies - your reply got lost in my inbox and I didn't see it until just now!

Yes, when --combineOutput is set to false, then you generate multiple files.  The filename should contain the book number and chapter number just as it does now.  Here are some examples:


Here is an example of combining the output into a single file:
```
User selects Mark 1, Mark 2, Mark 3, Mark 4
--combineOutput = true
single filename = Mark1-4.mp4
```

Here is an example of generating multiple files on output:
```
User selects Mark 1, Mark 2, Mark 3, Mark 4
--combineOutput = false
multiple files:
Mark1.mp4
Mark2.mp4
Mark3.mp4
Mark4.mp4
```

These are examples where the default filename is used.  In actuality, the UI lets the user pick a full file path, which includes the filename.  If the user does not take the default filename (say they pick "Bui.mp4" as the output file), then just append a number to each generated file e.g. Bui1.mp4 Bui2.mp4 ...

File: public/cli/lib/commands/convert.js [hunk 1]
```diff
@@ -410,69 +427,130 @@ function removeOutputFile(done) {
...
+        }));
+
+        // Combine videos
+        tasks.push(() => new Promise((next, fail) => {
+            let ffmpegExe = Options.ffmpegPath || "ffmpeg";
+            Log(`>> Combining videos`);
+
+            onProgress("Combining videos...", 100);
+            shell.exec(
+                `"${ffmpegExe}" -f concat -safe 0 -i "${outputListFile}" -c copy "${Options.output}"`,
```
Comment (megahirt) [inline public/cli/lib/commands/convert.js:544]: This seems like a good way to combine videos using FFMPEG.  Here is a document which talks about different approaches (for research purposes): https://trac.ffmpeg.org/wiki/Concatenate

File: public/cli/lib/commands/convert.js [hunk 2]
```diff
@@ -410,69 +427,130 @@ function removeOutputFile(done) {
...
+                (err) => {
```
Comment (megahirt) [inline public/cli/lib/commands/convert.js:545]: Would be nice to clean up the outputListFile after rendering (or upon failure).
```
fs.unlink(outputListFile);
```

File: public/karaoke.js [hunk 3]
```diff
@@ -104,4 +72,4 @@ async function execute({
 //     } catch (error) {
 //         console.error(error);
 //     }
-// })();
+// })();
```
Comment (megahirt) [inline public/karaoke.js:75]: need a newline at the end of the file

**Review by megahirt — COMMENTED** [review]
This is really good!  Just one or two small things.

File: public/cli/lib/commands/convert.js [hunk 4]
```diff
@@ -410,69 +427,130 @@ function removeOutputFile(done) {
...
+                        output,
+                        Log,
+                        framerateOut: Options.fps || 15,
+                        ffmpegPath: Options.ffmpegPath || null
+                    });
+                })
+            );
+    });
+
+    if (Options.combineOutput) {
```
Comment (megahirt) [inline public/cli/lib/commands/convert.js:521]:
```
if (Options.combinedOutput && outputFiles.length() > 1) {
```
Only combine output if more than one file is selected

File: public/cli/lib/commands/convert.js (hunk 2 above)
Comment (megahirt) [reply public/cli/lib/commands/convert.js:545]:
```
/// delete intermediate output files after combining them
outputFiles.map((f) => fs.unlink(f),
                (err) => {
                    return err ? fail(err): next();
                });
```

**Review by megahirt — CHANGES_REQUESTED** [review]

File: public/karaoke.js [hunk 5]
```diff
@@ -20,6 +20,11 @@ async function execute({
   onProgress,
 }) {
   try {
+hearThisFolder = [
```
Comment (megahirt) [inline public/karaoke.js]: obviously we will need to have more sensible defaults but this is fine for now

**Review by megahirt — APPROVED** [review]
AFAICT this looks good to me.  Great contribution, Pong!  We will do additional work once this gets merged into master.

**Pull request merged**